finality-aleph/src/data_io/data_interpreter.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{
+    data_io::{status_provider::ProposalStatusProvider, AlephData},
+    BlockHashNum, SessionBoundaries,
+};
+use async_trait::async_trait;
+use futures::channel::mpsc;
+use log::{debug, error, warn};
+use parking_lot::Mutex;
+use sp_runtime::traits::{Block as BlockT, NumberFor, One, Zero};
+use std::sync::Arc;
+
+/// Takes as input ordered `AlephData` from `AlephBFT` and pushes blocks that should be finalized
+/// to an output channel. The other end of the channel is held by the aggregator whose goal is to
+/// create multisignatures under the finalized blocks.
+pub struct OrderedDataInterpreter<B: BlockT, PSP: ProposalStatusProvider<B>> {
+    blocks_to_finalize_tx: mpsc::UnboundedSender<BlockHashNum<B>>,
+    proposal_status_provider: Arc<Mutex<PSP>>,
+    last_finalized_by_aleph: BlockHashNum<B>,
+    session_boundaries: SessionBoundaries<B>,
+}
+
+impl<B: BlockT, PSP: ProposalStatusProvider<B>> OrderedDataInterpreter<B, PSP> {
+    pub fn new(
+        blocks_to_finalize_tx: mpsc::UnboundedSender<BlockHashNum<B>>,
+        proposal_status_provider: Arc<Mutex<PSP>>,
+        session_boundaries: SessionBoundaries<B>,
+    ) -> Self {
+        let last_finalized_by_aleph = if session_boundaries.first_block() > NumberFor::<B>::zero() {
+            // We are in session > 0, we take the last block of previous session.
+            let last_prev_session_num = session_boundaries.first_block() - NumberFor::<B>::one();
+            proposal_status_provider
+                .lock()
+                .get_finalized_at(last_prev_session_num)
+                .expect("Previous session must have concluded")
+        } else {
+            // We are in session 0, we take the genesis block -- it is finalized by definition.
+            proposal_status_provider
+                .lock()
+                .get_finalized_at(NumberFor::<B>::zero())
+                .expect("Genesis block must be available")
+        };
+        OrderedDataInterpreter {
+            blocks_to_finalize_tx,
+            proposal_status_provider,
+            last_finalized_by_aleph,
+            session_boundaries,
+        }
+    }
+
+    fn block_to_finalize_from_data(&self, new_data: AlephData<B>) -> Option<BlockHashNum<B>> {
```

Comment:
This is now implemented up to some minor details.

Reply:
Eh, I haven't been clear enough and the "minor details" are what was imo quite important. >.<" To be precise I wanted this to be constructive, so returning `BlockHashNum` within the `Finalize` type was one of my main hopes.

Anyway the PR is now definitely in a reasonable state, so this would out of scope and at best a refactor for the future. Do you think I can add this ("Refactor `ProposalStatus` to be more constructive") as a task, or do you think doing it this way is a bad idea for some reason?